apps/api/src/domain/models/aggregate.entity.ts

```diff
@@ -0,0 +1,80 @@
+import { isInternalError } from '../../lib/errors/InternalError';
+import cloneToPlainObject from '../../lib/utilities/cloneToPlainObject';
+import { DeepPartial } from '../../types/DeepPartial';
+import { DTO } from '../../types/DTO';
+import { ResultOrError } from '../../types/ResultOrError';
+import { Valid } from '../domainModelValidators/Valid';
+import { AggregateCompositeIdentifier } from '../types/AggregateCompositeIdentifier';
+import { AggregateId } from '../types/AggregateId';
+import { AggregateType } from '../types/AggregateType';
+import { HasAggregateId } from '../types/HasAggregateId';
+import BaseDomainModel from './BaseDomainModel';
+import { BaseEvent } from './shared/events/base-event.entity';
+
+export abstract class Aggregate extends BaseDomainModel implements HasAggregateId {
+    /**
+     * We make this property optional so we don't need to specify it on existing data
+     * or test data. If it is not on a DTO, it will be set to [] in the constructor.
+     */
+    readonly eventHistory?: BaseEvent[];
+
+    readonly type: AggregateType;
+
+    readonly id: AggregateId;
+
+    constructor(dto: DTO<Aggregate>) {
+        super();
+
+        // This should only happen in the validation flow
+        if (!dto) return;
+
+        this.type = dto.type;
+
+        this.id = dto.id;
+
+        this.eventHistory = Array.isArray(dto.eventHistory)
+            ? cloneToPlainObject(dto.eventHistory)
+            : [];
+    }
+
+    getCompositeIdentifier = (): AggregateCompositeIdentifier => ({
+        type: this.type,
+        id: this.id,
+    });
+
+    protected safeClone<T extends Aggregate>(
+        this: T,
+        updateDto: DeepPartial<DTO<T>>
+    ): ResultOrError<T> {
+        const updatedInstance = this.clone<T>(updateDto);
+
+        const validationResult = updatedInstance.validateInvariants();
+
+        if (isInternalError(validationResult)) return validationResult;
+
+        return updatedInstance;
+    }
+
+    abstract validateInvariants(): ResultOrError<Valid>;
+
+    abstract getAvailableCommands(): string[];
+
+    /**
+     * The name of this method is a bit misleading. It merely adds an event
```

Comment:
Will all events be commands?  For the moment could this be something like `commandHistory` and `registerCommandInHistory`?  Not sure that either of those things make sense, especially because they won't help us if we do go to ES, but just trying some things out...

Reply:
No. It's not accurate to call an event a command. An event is a record of something that's already happened. I recall adding this comment and renaming the method based on feedback you had given. What's weird is that we update the state then record the event and it feels like the event should lead the way.